src/axom/core/Array.hpp

```diff
@@ -55,11 +88,152 @@ bool operator==(const Array<T>& lhs, const Array<T>& rhs);
+   */
+  template <typename... Args,
+            typename SFINAE = typename std::enable_if<sizeof...(Args) == DIM>::type>
+  T& operator()(Args... args)
+  {
+    IndexType indices[] = {static_cast<IndexType>(args)...};
+    IndexType idx =
+      std::inner_product(indices, indices + DIM, m_strides.begin(), 0);
+    // assert(inBounds(idx));
+    return asDerived().data()[idx];
+  }
+  /// \overload
+  template <typename... Args,
+            typename SFINAE = typename std::enable_if<sizeof...(Args) == DIM>::type>
+  const T& operator()(Args... args) const
+  {
+    IndexType indices[] = {static_cast<IndexType>(args)...};
+    IndexType idx =
+      std::inner_product(indices, indices + DIM, m_strides.begin(), 0);
+    // assert(inBounds(idx));
+    return asDerived().data()[idx];
+  }
+
+  /// \brief Swaps two ArrayImpls
+  friend void swap(ArrayImpl& lhs, ArrayImpl& rhs)
+  {
+    std::swap(lhs.m_dims, rhs.m_dims);
+    std::swap(lhs.m_strides, rhs.m_strides);
+  }
+
+  /// \brief Returns the dimensions of the Array
+  const std::array<IndexType, DIM>& shape() const { return m_dims; }
+
+  /// \brief Returns the strides of the Array
+  const std::array<IndexType, DIM>& strides() const { return m_strides; }
+
+private:
+  ArrayType& asDerived() { return static_cast<ArrayType&>(*this); }
+  const ArrayType& asDerived() const
+  {
+    return static_cast<const ArrayType&>(*this);
+  }
+
+  // Is it sufficient to just give them the raw arrays?
+  /// \brief The sizes (extents?) in each dimension
+  std::array<IndexType, DIM> m_dims;
+  /// \brief The strides in each dimension
+  std::array<IndexType, DIM> m_strides;
+};
+
+/// \brief Array implementation specific to 1D Arrays
+template <typename T>
+class ArrayImpl<T, 1>
+{
+public:
+  using ArrayType = Array<T, 1>;
+  ArrayImpl(IndexType = 0) { }
+
+  /*!
+   * \brief Push a value to the back of the array.
+   *
+   * \param [in] value the value to the back.
+   *
+   * \note Reallocation is done if the new size will exceed the capacity.
+   */
+  void push_back(const T& value);
+
+  /*!
+   * \brief Push a value to the back of the array.
+   *
+   * \param [in] value the value to move to the back.
+   *
+   * \note Reallocation is done if the new size will exceed the capacity.
+   */
+  void push_back(T&& value);
+
+  /*!
+   * \brief Inserts new element at the end of the Array.
+   *
+   * \param [in] args the arguments to forward to constructor of the element.
+   *
+   * \note Reallocation is done if the new size will exceed the capacity.
+   * \note The size increases by 1.
+   */
+  template <typename... Args>
+  void emplace_back(Args&&... args);
+
+  /// \brief Returns the dimensions of the Array
+  // FIXME: std::array is used for consistency with multidim case, should we just return the scalar?
+  // Double curly braces needed for C++11 prior to resolution of CWG issue 1720
+  std::array<IndexType, 1> shape() const { return {{asDerived().size()}}; }
```

Comment:
Minor: Is it weird for this to return a `std::array`?  
Perhaps it should be an `axom::StackArray`  (assuming we're keeping that) ?

Reply:
This will require some additional functionality in `StackArray` so I will log this in the main issue and come back to it later - before I start working on full GPU support most likely.